homeassistant/components/energy/sensor.py

```diff
@@ -0,0 +1,86 @@
+"""Helper sensor for calculating utility costs."""
+from __future__ import annotations
+
+from datetime import timedelta
+from typing import Any
+
+from homeassistant.components.sensor import (
+    ATTR_LAST_RESET,
+    DEVICE_CLASS_MONETARY,
+    SensorEntity,
+)
+from homeassistant.const import ATTR_NAME
+from homeassistant.core import callback
+from homeassistant.helpers.event import async_track_state_change_event
+import homeassistant.util.dt as dt_util
+
+# TODO: Spawn entities
+
+
+class EnergyCostSensor(SensorEntity):
+    """Calculate costs incurred by consuming energy."""
+
+    def __init__(
+        self,
+        currency: str,
+        energy_sensor: str,
+        energy_price_sensor: str,
+        hourly_price_sensor: str,
+    ) -> None:
+        """Initialize the sensor."""
+        super().__init__()
+
+        self._attr_device_class = DEVICE_CLASS_MONETARY
+        self._attr_unit_of_measurement = currency
+
+        self._energy_sensor_entity_id = energy_sensor
+        self._energy_price_entity_id = energy_price_sensor
+        self._hourly_price_entity_id = hourly_price_sensor
+
+    def _update_cost(self) -> None:
+        """Update incurred costs."""
+        # TODO: This will use the same prices for the entire duration, we should
+        # instead apply changed prices only after the price change.
+        energy_state = self.hass.states.get(self._energy_sensor_entity_id)
+        energy_price_state = self.hass.states.get(self._energy_sensor_entity_id)
+        hourly_price_state = self.hass.states.get(self._energy_sensor_entity_id)
+        if (
+            energy_state is None
+            or energy_price_state is None
+            or hourly_price_state is None
+            or ATTR_LAST_RESET not in energy_state.attributes
+        ):
```

Comment:
Do we also need to guard for unknown or unavailable states?

Reply:
Yes, I think so. Should we set the cost as unavailable too when that happens?

Reply:
Probably, yes.

Reply:
Current implementation we decided that when the consumption or price is not a float, we just don't process it. We will have the state be the accumulated value.